### PR TITLE
Flip

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
# PR Description

Provides partial flip, split and merge operations for half-edge data structures and triangulated surfaces.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: src/DGtal/shapes/TriangulatedSurface.h [hunk 1]
```diff
@@ -546,6 +550,77 @@ namespace DGtal
     */
     VertexRange allBoundaryVertices() const;
 
+    /**
+       @param a any arc.
+       @return either the four vertices of the two triangles bordering \a a, or the three vertices if this is an edge on the boundary.
+       @note O(1) operation
+       @note Order is, if arc a is (s,t), and a belongs to triangle (s,t,u) and opposite arc belongs to triangle (t,s,v): (t,u,s,v), (t,u,s), (t,s,v) 
+    */
+    VertexRange verticesAroundArc( const Arc a ) const;
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.h]:
```suggestion
    VertexRange verticesOfFacesAroundArc( const Arc a ) const;
```
This is a bit overloaded but less confusing that "verticesAroundArc" which would be read as "(the two) vertices of an arc"

File: src/DGtal/shapes/TriangulatedSurface.h [hunk 2]
```diff
@@ -546,6 +550,77 @@ namespace DGtal
...
+      
+    // ----------------------- Other services ---------------------------------
+  public:
+
+    /// An edge is (topologically) flippable iff: (1) it does not lie
+    /// on the boundary, (2) it is bordered by two triangles. (2) is
+    /// always true for a triangulated surface.
+    ///
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.h]: O(1) too, I guess

File: src/DGtal/shapes/TriangulatedSurface.h [hunk 3]
```diff
@@ -546,6 +550,77 @@ namespace DGtal
...
+    /// @return 'true' if the edge containing \a a is topologically flippable.
+    bool isFlippable( const Arc a ) const;
+
+    /// Flip the edge associated to arc \a a.
+    ///
+    /// @param a any valid arc.
+    ///
+    /// @pre the edge must be flippable, `isFlippable( a ) == true`
+    /// @see isFlippable
+    ///
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.h]: maybe you could add the complexities as well for these elementary "Euler" operations..

File: src/DGtal/shapes/TriangulatedSurface.h [hunk 4]
```diff
@@ -546,6 +550,77 @@ namespace DGtal
...
+    /// @return the index of the created vertex.
+    ///
+    /// @pre the edge must be flippable, `isFlippable( a ) == true`
+    /// @see isFlippable
+    Vertex split( const Arc a, const Point& data );
+
+    /// An edge is (topologically) mergeable iff: (1) it is bordered
+    /// by two triangles (for now), (2) there is no boundary vertex in
+    /// the two triangles bordering the edge (for now).
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.h]: what do you mean by the `for now` ?

File: src/DGtal/shapes/TriangulatedSurface.h (hunk 1 above)
Comment (dcoeurjo) [reply src/DGtal/shapes/TriangulatedSurface.h]: (head(a) tail(a) metthods)

File: src/DGtal/shapes/TriangulatedSurface.h [hunk 5]
```diff
@@ -546,6 +550,77 @@ namespace DGtal
...
+    ///
+    /// @param a any arc.
+    /// @return 'true' if the edge containing \a a is topologically mergeable.
+    bool isMergeable( const Arc a ) const;
+
+    /// Merges the edge specified by the half-edge \a i.
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.h]: there is no param "i"

File: src/DGtal/shapes/TriangulatedSurface.h [hunk 6]
```diff
@@ -546,6 +550,77 @@ namespace DGtal
...
+    /// @see isMergeable
+    /// @todo We could also merge boundary triangles.
+    ///
+    /// @note All the underlying data-structure is renumbered so that
+    /// vertex, edge, face and half-edge indices are consecutive. In
+    /// other words, since 1 vertex, 3 edges, 2 faces and 6 half-edges
+    /// have been deleted, then the last vertices, edges, faces,
+    /// half-edges have been renumbered into the freed indices. For
+    /// instance, the vertex `head(a)` does not exist anymore (but has
+    /// been replaced by a renumbered vertex).
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.h]: you should say that it is still O(1)

File: src/DGtal/shapes/TriangulatedSurface.ih [hunk 7]
```diff
@@ -267,7 +267,8 @@ typename DGtal::TriangulatedSurface<TPoint>::Arc
 DGtal::TriangulatedSurface<TPoint>::arc
 ( const Vertex & t, const Vertex & h ) const
 {
-  return myHEDS.halfEdgeIndexFromArc( t, h );
+  //return myHEDS.halfEdgeIndexFromArc( t, h );
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.ih]: clean

File: src/DGtal/shapes/TriangulatedSurface.ih [hunk 8]
```diff
@@ -349,6 +350,40 @@ DGtal::TriangulatedSurface<TPoint>::isArcBoundary( const Arc& v ) const
   return INVALID_FACE == myHEDS.halfEdge( v ).face;
 }
 
+//-----------------------------------------------------------------------------
+template <typename TPoint>
+inline
+typename DGtal::TriangulatedSurface<TPoint>::VertexRange
+DGtal::TriangulatedSurface<TPoint>::verticesAroundArc( Arc a ) const
```
Comment (dcoeurjo) [inline src/DGtal/shapes/TriangulatedSurface.ih]: cf `.h`

File: src/DGtal/shapes/doc/moduleHalfEdgeMesh.dox [hunk 9]
```diff
@@ -162,6 +162,32 @@ It is worthy to note the following elements in this representation:
...
+- HalfEdgeDataStructure::isMergeable tells if some edge is mergeable,
+  i.e. (1) it is bordered by two triangles, (2) there is no boundary
+  vertex in the two triangles bordering the edge;
+- HalfEdgeDataStructure::merge merges the given edge and the two
+  bordering triangles, leaving only one vertex and two edges, only if
+  the edge is mergeable.
+
+There is also a method HalfEdgeDataStructure::isValid that performs a
+lot of checks on the validity of the data structure.
+
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleHalfEdgeMesh.dox]: add  something saying that all these operations are O(1)

Comment (dcoeurjo) [reply src/DGtal/shapes/doc/moduleHalfEdgeMesh.dox]: if they are

Comment (dcoeurjo) [reply src/DGtal/shapes/doc/moduleHalfEdgeMesh.dox]: well.. there are not for HDS.. but they are for Triangulated surfaces.

File: src/DGtal/topology/HalfEdgeDataStructure.h [hunk 10]
```diff
@@ -545,6 +579,788 @@ namespace DGtal
...
+			<< " nbV1=" << nbV1
+			<< " nbV2=" << nbV2
+			<< " nbV3=" << nbV3
+			<< " nbV4=" << nbV4 << std::endl
+			<< " new_nbV1=" << new_nbV1
+			<< " new_nbV3=" << new_nbV3
+			<< " new_nbV4=" << new_nbV4 << std::endl;
+
+      // Renumbering of 1 vertex, 3 edges, 2 faces, 6 half-edges
+      // std::cout << "Renumbering vertex" << v2 << "/" << nbVertices() << std::endl;
```
Comment (dcoeurjo) [inline src/DGtal/topology/HalfEdgeDataStructure.h]: clean of verbose mode

File: src/DGtal/topology/HalfEdgeDataStructure.h [hunk 11]
```diff
@@ -545,6 +579,788 @@ namespace DGtal
...
+      renumberVertex( v2, update_arc2index );
+      std::array< EdgeIndex, 3 > E = { e1, e2, e3 };
+      std::sort( E.begin(), E.end(), std::greater<EdgeIndex>() );
+      for ( Index e : E ) {
+	// std::cout << "Renumbering edge" << e  << "/" << nbEdges() << std::endl;
```
Comment (dcoeurjo) [inline src/DGtal/topology/HalfEdgeDataStructure.h]: same

**Comment by dcoeurjo** [comment]
You would also need to update the changelog conflict.

File: src/DGtal/shapes/TriangulatedSurface.h (hunk 1 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.h]: Ok. I've changed as you said.

File: src/DGtal/shapes/TriangulatedSurface.h (hunk 2 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.h]: yes. I've updated the comments

File: src/DGtal/shapes/TriangulatedSurface.h (hunk 3 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.h]: done

File: src/DGtal/shapes/TriangulatedSurface.h (hunk 4 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.h]: I remove these comments.

File: src/DGtal/shapes/TriangulatedSurface.h (hunk 5 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.h]: you're right

File: src/DGtal/shapes/TriangulatedSurface.h (hunk 6 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.h]: yes. done.

File: src/DGtal/shapes/TriangulatedSurface.ih (hunk 7 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.ih]: done

File: src/DGtal/shapes/TriangulatedSurface.ih (hunk 8 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/TriangulatedSurface.ih]: done

File: src/DGtal/shapes/doc/moduleHalfEdgeMesh.dox (hunk 9 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/doc/moduleHalfEdgeMesh.dox]: I've updated the doc

File: src/DGtal/topology/HalfEdgeDataStructure.h (hunk 11 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/topology/HalfEdgeDataStructure.h]: done

**Comment by dcoeurjo** [comment]
Merging

**Pull request merged**